### PR TITLE
Code design and legacy cleanup of datatypes and metadata.

### PR DESCRIPTION
**Pull request by jmchilton**
- Eliminates legacy entry points for setting metadata that were added with 15.03 only for running jobs. This will cause jobs running during an upgrade from pre 15.03 to post 16.04 to fail, they were going to fail anyway since that upgrade is going to be tricky. 76f898e 31f9aae
- General style and import fixes (eliminates many relative imports in galaxy.datatypes). cc4e9e1 a6ae9e1
- A series of refactorings that eliminate any import-time dependency in `galaxy.model` on `galaxy.datatypes`, establishing a strict ordering for these modules and fixing #1552.
- Improved use of the global datatypes_registry in `galaxy.model` - don't auto-create a default registry used just for unit tests (should speed up jobs).
- Eliminate the import-time dependency of in non-converter code of `galaxy.datatypes` on `galaxy.tools`. There are still "problems" with the converters loading too much of `galaxy.tools` and a runtime import of `galaxy.tools.utils.maf_utilities` that I'm not happy with - but the import situation should be much better now and we've established the principle that `galaxy.datatypes` shouldn't depend on `galaxy.tools` in the abstract I think.

### CODE REVIEW

**Comment by nsoranzo** [comment]
@jmchilton You may want to check if after this the imports of `galaxy.model` in `tools/` can be removed/fixed:

```
$ ack 'galaxy.model' galaxyproject_galaxy/tools/
galaxyproject_galaxy/tools/data_source/data_source.py
10:# need to import TOOL_PROVIDED_JOB_METADATA_FILE (which will import galaxy.model) before sniff to resolve a circular import between galaxy.datatypes.registry and galaxy.model

galaxyproject_galaxy/tools/data_source/upload.py
18:import galaxy.model  # noqa

galaxyproject_galaxy/tools/genomespace/genomespace_importer.py
13:import galaxy.model # need to import model before sniff to resolve a circular import dependency

galaxyproject_galaxy/tools/genomespace/genomespace_file_browser.py
12:import galaxy.model # need to import model before sniff to resolve a circular import dependency
```

**Comment by jmchilton** [comment]
@nsoranzo - Good call, I am almost certain those will not longer be needed.

**Comment by jmchilton** [comment]
@nsoranzo https://github.com/galaxyproject/galaxy/commit/499357e508035e7c9ffb5d711e0f355e960e1e05.

I didn't modify the genomespace stuff, I just updated the comments. I don't have a good test apparatus for changes to those files.

File: lib/galaxy/datatypes/sniff.py [hunk 1]
```diff
@@ -253,75 +254,75 @@ def is_column_based( fname, sep='\t', skip=0, is_multi_byte=False ):
     return True
 
 
-def guess_ext( fname, sniff_order=None, is_multi_byte=False ):
+def guess_ext( fname, sniff_order, is_multi_byte=False ):
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sniff.py:257]: `tools/data_source/upload.py` contains this line:

```
ext = sniff.guess_ext( dataset.path, is_multi_byte=True )
```

which will break after this change.

Comment (jmchilton) [inline lib/galaxy/datatypes/sniff.py:257]: :scream_cat: How did none of the tests catch that, I'll look into this - thanks @nsoranzo.

Comment (jmchilton) [inline lib/galaxy/datatypes/sniff.py:257]: I believe I have fixed this, we need testing of the multi-byte char stuff - unfortunately I don't have an even rudimentary understanding of the underlying issues or why those branches exist in the code.

**Comment by nsoranzo** [comment]
:+1:

**Comment by martenson** [comment]
thanks @jmchilton & @nsoranzo

**Pull request merged**